### PR TITLE
Add basic stubs for native extension

### PR DESCRIPTION
**Pull request by daskol**

### CODE REVIEW

File: wrappers/python/nntile/__init__.py [hunk 1]
```diff
@@ -11,5 +11,8 @@
 # @author Aleksandr Mikhalev
 # @date 2023-02-22
```
Comment (Muxas) [inline wrappers/python/nntile/__init__.py:12]: Add yourself to the list of authors of the file and update data

File: wrappers/python/nntile/nntile_core/__init__.pyi [hunk 2]
```diff
@@ -0,0 +1,6 @@
+class TransOp:
+    """Transposition operation."""
```
Comment (Muxas) [inline wrappers/python/nntile/nntile_core/__init__.pyi:2]: This is not an operation, but a "Transposition flag data type"

File: wrappers/python/nntile/nntile_core/starpu.pyi [hunk 3]
```diff
@@ -0,0 +1,18 @@
+class Config:
```
Comment (Muxas) [inline wrappers/python/nntile/nntile_core/starpu.pyi:1]: Add a header to the file. You can copy header from another Python file.

File: wrappers/python/nntile/nntile_core/tensor.pyi [hunk 4]
```diff
@@ -0,0 +1,55 @@
+from collections.abc import Buffer
```
Comment (Muxas) [inline wrappers/python/nntile/nntile_core/tensor.pyi:1]: Add a header to the file

File: wrappers/python/nntile/nntile_core/tensor.pyi [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+    def __init__(self, shape: Sequence[int], basetile_shape: Sequence[int]):
+        ...
+
+    def __repr__(self) -> str: ...
+
+    @property
+    def basetile_shape(self) -> tuple[int, ...]: ...
+
+    @property
+    def grid(self) -> Tile: ...
```
Comment (Muxas) [inline wrappers/python/nntile/nntile_core/tensor.pyi:18]: Actual type of the returned value is "TileTraits"

Comment (Muxas) [reply wrappers/python/nntile/nntile_core/tensor.pyi:18]: Seems like this Python include file was not run by a Python. How can we test such files?

File: wrappers/python/nntile/nntile_core/tensor.pyi [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+class Tensor(TensorTraits):
+    """Many-dimensional tensor, presented by a set of subtensors (tiles)."""
+
+    def __init__(self, traits: TensorTraits, distribution: Sequence[int, ...],
+                 last_tag: int): ...
+
+    @property
+    def next_tag(self) -> int: ...
+
+    def from_array(self, array: Buffer) -> None: ...
```
Comment (Muxas) [inline wrappers/python/nntile/nntile_core/tensor.pyi:33]: Actual data type is not presented here neither for Tensor nor for Buffer. How does it work in such cases? Do we have to store additional runtime member "dtype" to derive actual data types?

File: wrappers/python/nntile/nntile_core/tile.pyi [hunk 7]
```diff
@@ -0,0 +1,42 @@
+from collections.abc import Buffer
```
Comment (Muxas) [inline wrappers/python/nntile/nntile_core/tile.pyi:1]: Add header

File: wrappers/python/nntile/nntile_core/tile.pyi [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+    def shape(self) -> tuple[int, ...]: ...
+
+    @property
+    def stride(self) -> tuple[int, ...]: ...
+
+    @property
+    def nelems(self) -> int: ...
+
+    def linear_to_index(self, linear_offset: int) -> tuple[int, ...]: ...
+    def index_to_linear(self, linear_offset: tuple[int, ...]) -> int: ...
```
Comment (Muxas) [inline wrappers/python/nntile/nntile_core/tile.pyi:23]: parameter is not "linear_offset", it is an "index_offset"

File: wrappers/python/nntile/nntile_core/tensor.pyi [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+    def to_array(self) -> Buffer: ...
+
+    def get_tile(self, index: Sequence[int, ...]) -> Tile: ...
```
Comment (Muxas) [inline wrappers/python/nntile/nntile_core/tensor.pyi:36]: There shall be another overloaded function get_tile with "index: int"

File: wrappers/python/nntile/nntile_core/tile.pyi [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+
+
+class Tile(TileTraits):
+    """Many-dimensional tensor, stored contiguously in a Fortran order."""
+
+    def __init__(self, traits: TileTraits): ...
+    def from_array(self, array: Buffer): ...
+    def to_array(self, array: Buffer): ...
```
Comment (Muxas) [inline wrappers/python/nntile/nntile_core/tile.pyi:31]: Just like with the Tensor case, it is preferable to convert parameter array into an output

File: wrappers/python/nntile/nntile_core/tile.pyi [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+    def unregister(self) -> None: ...
+
+
+def tile_from_array(tile: Tile, array: Buffer): ...
+def tile_to_array(tile: Tile, array: Buffer): ...
```
Comment (Muxas) [inline wrappers/python/nntile/nntile_core/tile.pyi:36]: Make array a return value

File: wrappers/python/nntile/nntile_core/tensor.pyi [hunk 12]
```diff
@@ -0,0 +1,55 @@
...
+    def set_reduction_add(self) -> None: ...
+    def set_reduction_hypot(self) -> None: ...
+    def set_reduction_maxsumexp(self) -> None: ...
+
+    def unregister(self) -> None: ...
+    def wont_use(self) -> None: ...
+
+
+def tensor_from_array(tensor: Tensor, array: Buffer): ...
+def tensor_to_array(tensor: Tensor, array: Buffer): ...
```
Comment (Muxas) [inline wrappers/python/nntile/nntile_core/tensor.pyi:49]: Convert array parameter into output value

**Review by Muxas — COMMENTED** [review]
This is just a draft. Is it really used in some way by Python?

File: wrappers/python/nntile/nntile_core/starpu.pyi (hunk 3 above)
Comment (daskol) [reply wrappers/python/nntile/nntile_core/starpu.pyi:1]: Why?

File: wrappers/python/nntile/nntile_core/tensor.pyi (hunk 4 above)
Comment (daskol) [reply wrappers/python/nntile/nntile_core/tensor.pyi:1]: Why?

File: wrappers/python/nntile/nntile_core/tensor.pyi (hunk 12 above)
Comment (daskol) [reply wrappers/python/nntile/nntile_core/tensor.pyi:49]: Nope, it does not return anything.

File: wrappers/python/nntile/nntile_core/tile.pyi (hunk 11 above)
Comment (daskol) [reply wrappers/python/nntile/nntile_core/tile.pyi:36]: It returns nothing as well.

File: wrappers/python/nntile/nntile_core/tile.pyi (hunk 10 above)
Comment (daskol) [reply wrappers/python/nntile/nntile_core/tile.pyi:31]: Well, it return nothing...

File: wrappers/python/nntile/nntile_core/tensor.pyi (hunk 9 above)
Comment (daskol) [reply wrappers/python/nntile/nntile_core/tensor.pyi:36]: Ah, `Index` is an alias for `int64_t`. We should avoid doing this.

File: wrappers/python/nntile/nntile_core/tensor.pyi (hunk 5 above)
Comment (daskol) [reply wrappers/python/nntile/nntile_core/tensor.pyi:18]: The purpose of these files is actually testing, lol. Stub files define types then type checker like `mypy` or `pylint` verifies that type system is consistent.

File: wrappers/python/nntile/nntile_core/tensor.pyi (hunk 6 above)
Comment (daskol) [reply wrappers/python/nntile/nntile_core/tensor.pyi:33]: Well, `collections.abc.Buffer` is dependent type but at the moment it is too much. There is a lot to do in order to `mypy` does not cry.

File: wrappers/python/nntile/__init__.py (hunk 1 above)
Comment (daskol) [reply wrappers/python/nntile/__init__.py:12]: I think that `git` does it better than human.

**Comment by daskol** [comment]
> This is just a draft. Is it really used in some way by Python?

No, it is not a draft. Stubs for opts and codelets are still missing but this is the next step.

File: wrappers/python/nntile/nntile_core/starpu.pyi (hunk 3 above)
Comment (Muxas) [reply wrappers/python/nntile/nntile_core/starpu.pyi:1]: A header like
```
# @copyright (c) 2022-2023 Skolkovo Institute of Science and Technology
#                           (Skoltech). All rights reserved.
#
# NNTile is software framework for fast training of big neural networks on
# distributed-memory heterogeneous systems based on StarPU runtime system.
#
# @file wrappers/python/nntile/pipeline.py
# TRaining pipeline of NNTile Python package
#
# @version 1.0.0
# @author Aleksandr Mikhalev
# @author Aleksandr Katrutsa
# @date 2023-09-20
```
tells a reader that the file is a part of NNTile project with its location. Sometimes files are just copied from a project to project without any `git` history. Probably, list of authors and date of last update are an overkill, but altogether such an information helps external collaborators.

File: wrappers/python/nntile/nntile_core/tensor.pyi (hunk 5 above)
Comment (Muxas) [reply wrappers/python/nntile/nntile_core/tensor.pyi]: If purpose of such files is testing then why there was no testing? Running the code would show the error.

File: wrappers/python/nntile/nntile_core/tensor.pyi [hunk 13]
```diff
@@ -0,0 +1,55 @@
...
+    """Many-dimensional tensor, presented by a set of subtensors (tiles)."""
+
+    def __init__(self, traits: TensorTraits, distribution: Sequence[int],
+                 last_tag: int): ...
+
+    @property
+    def next_tag(self) -> int: ...
+
+    def from_array(self, array: Buffer) -> None: ...
+    def to_array(self) -> Buffer: ...
```
Comment (Muxas) [inline wrappers/python/nntile/nntile_core/tensor.pyi:34]: This `to_array` function returns value, while other `to_array` functions do not return anything. This is why I was asking to convert other `to_array` functions.

File: wrappers/python/nntile/nntile_core/__init__.pyi [hunk 14]
```diff
@@ -0,0 +1,6 @@
+class TransOp:
```
Comment (Muxas) [inline wrappers/python/nntile/nntile_core/__init__.pyi:1]: Insert header

**Comment by Muxas** [comment]
> No, it is not a draft. Stubs for opts and codelets are still missing but this is the next step.

This is not a draft and it contains mistakes. So, it was not actually tested. Draft means it is not yet ready to be run and tested.

**Pull request closed without merging**